File: core/src/main/java/javabot/dao/LogsDao.java
package javabot.dao;

import javabot.Seen;
import javabot.model.Logs;
import javabot.model.Logs.Type;
import javabot.model.criteria.LogsCriteria;
import org.pircbotx.Channel;
import org.pircbotx.User;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Collections;
import java.util.List;

public class LogsDao extends BaseDao<Logs> {
    public static final Logger LOG = LoggerFactory.getLogger(LogsDao.class);
    @Inject
    public ConfigDao dao;
    @Inject
    public ChannelDao channelDao;

    public LogsDao() {
        super(Logs.class);
    }

    public void logMessage(final Type type, final Channel channel, final User user, final String message) {
        final Logs logMessage = new Logs();
        logMessage.setType(type);
        if (user != null) {
            logMessage.setNick(user.getNick());
        }
        if (channel != null) {
            logMessage.setChannel(channel.getName().toLowerCase());
        }
        logMessage.setMessage(message);
        logMessage.setUpdated(LocalDateTime.now());
        save(logMessage);
    }

    public boolean isSeen(final String channel, final String nick) {
        return getSeen(channel, nick) != null;
    }

    public Seen getSeen(final String channel, final String nick) {
        LogsCriteria criteria = new LogsCriteria(ds);
        criteria.upperNick().equal(nick.toUpperCase());
        criteria.channel().equal(channel);
        criteria.updated().order(false);
        Logs logs = criteria.query().get();
        return logs != null ? new Seen(logs.getChannel(), logs.getMessage(), logs.getNick(), logs.getUpdated()) : null;
    }

    private List<Logs> dailyLog(String channelName, LocalDateTime date, Boolean logged) {
        List<Logs> list = null;
        if (logged) {
            LocalDate start = date == null
                              ? LocalDate.now()
                              : date.toLocalDate();
            LocalDate tomorrow = start.plusDays(1);
            LogsCriteria criteria = new LogsCriteria(ds);
            criteria.channel(channelName);
            LocalDateTime nextMidnight = tomorrow.atStartOfDay();
            LocalDateTime lastMidnight = start.atStartOfDay();
            criteria.and(
                            criteria.updated().lessThanOrEq(nextMidnight),
                            criteria.updated().greaterThanOrEq(lastMidnight)
                        );
            list = criteria.query().asList();
        }
        return list;
    }

    public List<Logs> findByChannel(String name, LocalDateTime date, Boolean showAll) {
        javabot.model.Channel channel = channelDao.get(name);
        if (channel != null && (showAll || channel.getLogged())) {
            return dailyLog(name, date, showAll || channel.getLogged());
        } else {
            return Collections.emptyList();
        }
    }

    public void deleteAllForChannel(final String channel) {
        LogsCriteria criteria = new LogsCriteria(ds);
        criteria.channel(channel);
        ds.delete(criteria.query());
    }
}

File: src/main/kotlin/javabot/dao/LogsDao.kt
package javabot.dao

import com.google.inject.Inject
import dev.morphia.Datastore
import dev.morphia.DeleteOptions
import dev.morphia.query.FindOptions
import dev.morphia.query.Sort
import dev.morphia.query.filters.Filters.eq
import dev.morphia.query.filters.Filters.gte
import dev.morphia.query.filters.Filters.lte
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.Locale
import javabot.Seen
import javabot.model.Channel
import javabot.model.JavabotUser
import javabot.model.Logs
import javabot.model.Logs.Type
import org.slf4j.Logger
import org.slf4j.LoggerFactory

class LogsDao @Inject constructor(ds: Datastore, var dao: ConfigDao, var channelDao: ChannelDao) :
    BaseDao<Logs>(ds, Logs::class.java) {

    companion object {
        val LOG: Logger = LoggerFactory.getLogger(LogsDao::class.java)
    }

    fun logMessage(type: Type, channel: Channel?, user: JavabotUser?, message: String) {
        save(Logs(user?.nick, message, type, channel?.name))
    }

    fun isSeen(channel: String, nick: String): Boolean {
        return getSeen(channel, nick) != null
    }

    fun getSeen(channel: String, nick: String): Seen? {
        val first =
            ds.find(Logs::class.java)
                .filter(
                    eq("upperNick", nick.uppercase(Locale.getDefault())),
                    eq("channel", channel),
                )
                .first(FindOptions().sort(Sort.descending("updated")))

        return first?.let { Seen(it.channel!!, it.message, it.nick, it.updated) }
    }

    private fun dailyLog(channelName: String, date: LocalDateTime?, logged: Boolean) =
        if (logged) listOf()
        else {
            val start = if (date == null) LocalDate.now() else date.toLocalDate()
            val tomorrow = start.plusDays(1)
            val nextMidnight = tomorrow.atStartOfDay()
            val lastMidnight = start.atStartOfDay()
            ds.find<Logs>(Logs::class.java)
                .filter(
                    eq("channel", channelName),
                    lte("updated", nextMidnight),
                    gte("updated", lastMidnight),
                )
                .iterator(FindOptions().sort(Sort.ascending("updated")))
                .toList()
        }

    fun findByChannel(name: String, date: LocalDateTime, showAll: Boolean): List<Logs> {
        val channel = channelDao.get(name)
        if (channel != null && (showAll || channel.logged)) {
            return dailyLog(name, date, showAll || channel.logged)
        } else {
            return emptyList()
        }
    }

    fun deleteAllForChannel(channel: String) {
        ds.find(Logs::class.java).filter(eq("channel", channel)).delete(DeleteOptions().multi(true))
    }
}
